keystoremanager/keystoremanager.go

```diff
@@ -155,79 +162,127 @@ attempts to validate the certificate by first checking for an exact match on
 the certificate store, and subsequently trying to find a valid chain on the
 caStore.
 
-When this is being used with a notary repository, the dnsName parameter should
-be the GUN associated with the repository.
-
-Example TUF Content for root role:
-"roles" : {
-  "root" : {
-    "threshold" : 1,
-      "keyids" : [
-        "e6da5c303d572712a086e669ecd4df7b785adfc844e0c9a7b1f21a7dfc477a38"
-      ]
-  },
- ...
-}
+Currently this method operates on a Trust On First Use (TOFU) model: if we
+have never seen a certificate for a particular CN, we trust it. If later we see
+a different certificate for that certificate, we return an ErrValidationFailed error.
 
-Example TUF Content for root key:
-"e6da5c303d572712a086e669ecd4df7b785adfc844e0c9a7b1f21a7dfc477a38" : {
-	"keytype" : "RSA",
-	"keyval" : {
-	  "private" : "",
-	  "public" : "Base64-encoded, PEM encoded x509 Certificate"
-	}
-}
+Note that since we only allow trust data to be downloaded over an HTTPS channel
+we are using the current web-of-trust to validate the first download of the certificate
+adding an extra layer of security over the normal (SSH style) trust model.
+We shall call this: TOFUS.
+
+ValidateRoot also supports root key rotation, trusting a new certificate that has
+been included in the roots.json, and removing trust in the old one.
 */
 func (km *KeyStoreManager) ValidateRoot(root *data.Signed, dnsName string) error {
-	rootSigned := &data.Root{}
-	err := json.Unmarshal(root.Signed, rootSigned)
+	logrus.Debugf("entered ValidateRoot with dns: %s", dnsName)
+	rootSigned, err := data.RootFromSigned(root)
 	if err != nil {
 		return err
 	}
 
-	certs := make(map[string]*data.PublicKey)
-	for _, keyID := range rootSigned.Roles["root"].KeyIDs {
-		// TODO(dlaw): currently assuming only one cert contained in
-		// public key entry. Need to fix when we want to pass in chains.
-		k, _ := pem.Decode([]byte(rootSigned.Keys[keyID].Public()))
-		decodedCerts, err := x509.ParseCertificates(k.Bytes)
+	// Iterate over every keyID for the root role inside of roots.json
+	validKeys := make(map[string]*data.PublicKey)
+	allCerts := make(map[string]*x509.Certificate)
+	logrus.Debugf("found the following root keys in roots.json: %v", rootSigned.Signed.Keys)
+	for _, keyID := range rootSigned.Signed.Roles["root"].KeyIDs {
```

Comment:
Is the "root" entry guaranteed to exist at all times when this function is called? It might be a good idea to check anyway.